pkg/database/database.go

```diff
@@ -71,6 +73,22 @@ func NewMasterKeyAuthorizer(ctx context.Context, log *logrus.Entry, token azcore
 	return cosmosdb.NewMasterKeyAuthorizer(getDatabaseKey(keys, log))
 }
 
+func NewTokenAuthorizer(ctx context.Context, log *logrus.Entry, cred azcore.TokenCredential, databaseAccountName string, scopes []string) (cosmosdb.Authorizer, error) {
```

Comment:
I think there's a problem inherently with this function.  

The token we request from AAD with the scope is a short-lived credential (~12 hrs max I believe).  That means that at some point, the token will expire both from AAD MSI perspective and cosmosdb perspective.  As a result, we may need some refreshing of the actual database client in all places where it's used. 

Previously, we used the primary database keys which never expire but are rotatable.  That's why we didn't need to implement any database client reconstruction due to token expiration.  Now we more than likely do. 

The best way to test this would be to run an RP for a long time (or get the expiration of the token from this function) and see what happens when you get past the expiration time.

Reply:
The azcore.TokenCredential, it is indeed an object for tokens but it does not statically contains the token. It has a `getToken()` method which gives a refreshed token every time we call the method.
There is a token refresh logic in place in this go-cosmosdb code in this [PR](https://github.com/jewzaam/go-cosmosdb/pull/22) which cache the token. Every time we make a request, it is check is the current token is expired or 5 mins left to expire, if so it calls `getToken()` to get an updated token else use the cached ones.

Reply:
Is it possible on quieter regions that we might not have a request for over 12 hours? Would that have any effect?

Reply:
No, because everything a request goes to cosmosdb, there's a method `Authorize()` which gets called, and this method has the logic to acquire the token if the token is expired. The same logic which we were discussing lately.